parsl/executors/workqueue/executor.py

```diff
@@ -0,0 +1,410 @@
+
+import os
+import pickle
+
+from ipyparallel.serialize import pack_apply_message
+
+from parsl.executors.errors import *
+from parsl.executors.base import ParslExecutor
+from parsl.data_provider.files import File
+
+from work_queue import *
+
+logger = logging.getLogger(__name__)
+
+
+def WorkQueueThread(tasks={},
+                    task_queue=deque,
+                    queue_lock=threading.Lock(),
+                    tasks_lock=threading.Lock(),
+                    port=50000,
+                    launch_cmd=None,
+                    data_dir=".",
+                    log_dir=None,
+                    full=False,
+                    password=None,
+                    password_file=None,
+                    project_name=None,
+                    env=None):
+
+    logger.debug("Starting WorkQueue Master Thread")
+
+    wq_to_parsl = {}
+    if log_dir is not None:
+        wq_debug_log = os.path.join(log_dir, "debug")
+        cctools_debug_flags_set("all")
+        cctools_debug_config_file(wq_debug_log)
+
+    logger.debug("Creating Workqueue Object")
+    try:
+        q = WorkQueue(port)
+    except Exception as e:
+        logger.error("Unable to create Workqueue object: {}", format(e))
+        raise e
+
+    if project_name:
+        q.specify_name(project_name)
+
+    if password:
+        q.specify_password(password)
+    elif password_file:
+        q.specify_password_file(password_file)
+
+    # Only write Logs when the log_dir is specified, which is most likely always will be
+    if log_dir is not None:
+        wq_master_log = os.path.join(log_dir, "master_log")
+        wq_trans_log = os.path.join(log_dir, "transaction_log")
+        if not full:
+            wq_resource_log = os.path.join(log_dir, "resource_logs")
+            q.enable_monitoring_full(dirname=wq_resource_log)
+
+        q.specify_log(wq_master_log)
+        q.specify_transactions_log(wq_trans_log)
+
+    continue_running = True
+    while(continue_running):
+        # Monitor the Task Queue
+
+        place_holder_queue = []
+
+        queue_lock.acquire()
+        queue_len = len(task_queue)
+        while queue_len > 0:
+            place_holder_queue.append(task_queue.pop())
+            queue_len -= 1
+        queue_lock.release()
+
+        # Submit Tasks
+        for item in place_holder_queue:
+            parsl_id = item["task_id"]
+            if parsl_id < 0:
+                continue_running = False
+                break
+            function_data_loc = item["data_loc"]
+            function_result_loc = item["result_loc"]
+            function_result_loc_remote = function_result_loc.split("/")[-1]
+            function_data_loc_remote = function_data_loc.split("/")[-1]
+
+            input_files = item["input_files"]
+            output_files = item["output_files"]
+
+            # TODO Make this general
+            full_script_name = "/afs/crc.nd.edu/user/a/alitteke/parsl/parsl/executors/workqueue/workqueue_worker.py"
```

Comment:
The high throughput executor launches a script "process_worker_pool.py". It assumes that that script is in the current $PATH so can be launched without an explicit path - for example, we require users to have setup their virtualenv/conda environment in their job submission.

The install process makes that script available in the current path by placing it in the `script` block in `setup.py`.

Can your worker environment make the same assumption? If so, that might be a way to get rid of this path.

Reply:
I've pushed fixes addressing this below.

Reply:
I suggest the most foolproof way to go here is to have the WQ executor ship the script as an input file dependency from the Parsl install into the working directory of the task, then execute "./script".

Reply:
We need to somehow guarantee that the parsl-client side and worker-side environments match. Otherwise we run into deserialization issues from python version mismatches, and missing library issues. If the assumption is that the worker-side env is a clean slate, we'll have to ship the entire env (conda tarball ?) or send in a script that will setup the env. In parsl we have a `worker_init` string that's used to execute an arbitrary string on the worker side to setup the worker env, will something like that work here ? Is there some option I can pass to the work_queue_worker so that the env is inherited ?

Reply:
There is an `init_command` argument string for the executor that can be specified to run something before `work_queue_worker.py` is run.  You can also specify environmental variables for the worker with the `env` argument on the executor